### DISCUSSION TITLE
Set the Mantine theme object dynamically

### DISCUSSION THREAD

**Discussion by m-haris-n** [mantinedev/mantine#2310] [Q&A] [answered]
Greetings, everyone.
I need some help in a React app I'm making.
I want to populate my mantine theme colors by getting them from a backend api.
The thing is, I have tried it and it works if I don't use those dynamically created theme.
But the moment I use those theme colors in my components, it says that the color I'm referencing does not exist (since the theme object does not get updated in the first render, but in the second rerender)
Can someone please guide me about it?

**Comment by m-haris-n** [comment]
This is my code in the main App.js file, in a useEffect hook:

`const colors = getColors();`
`    colors.then((res) => {`
`      const colorData = res.data;`
`      setTheme({`
`        ...theme,`
`        colors: {`
`          base: [colorData.base, colorData.baseLight, colorData.baseDark],`
`          accent: [`
`            colorData.accent,`
`            colorData.accentLight,`
`            colorData.accentDark,`
`          ],`
`          neutral: [`
`            colorData.neutralDarkest,`
`            colorData.neutralDarker,`
`            colorData.neutralDark,`
`            colorData.neutral,`
`            colorData.neutralLight,`
`            colorData.neutralLighter,`
`            colorData.neutralLightest,`
`          ],`
`        },`
`        primaryColor: base,`
`      });`
`    });`

Now after doing it, if I console log it, it will get updated perfectly.
It will first console log the empty object, and in the second render, it will get updated.
But it doesn't work when I use the color I set and git an `ReferenceError: base is not defined`

**Comment by rtivital** [comment, marked as answer]
Store the theme object in state, initialize it with fallback values for colors that you want to add, then when you've received the data from the API update your state.

**Reply by m-haris-n** [reply]
Thanks a lot! That actually works

**Comment by bastiengrignon** [comment]
Hi, 
I'm reposting on this thread because I need to change the theme dynamically to update theme colors after an api call. I did it using **Redux** and storing my default theme object in the **initialState** of my slice but then when adding Mantine `globalStyles` in the object I got the following error in the console `A non-serializable value was detected in the state, in the path: ...`

I'll like to know if there is another solution to change and store theme dynamically whitout this error :thinking:

**Reply by m-haris-n** [reply]
can you show the code snippet and the error?

**Comment by bastiengrignon** [comment]
> can you show the code snippet and the error?
@m-haris-n Sure ! 

Here is my theme object : 
```js
export const theme = {
  globalStyles: () => ({
    'svg > circle': {
      transition: '0.5s all',
    }
  }),
  cursorType: 'pointer',
  spacing: {
    xxl: '2rem',
  },
};
```
Which is used as my initial state in a **redux slice** 
And this produce this error in the console: 
```shell
A non-serializable value was detected in the state, in the path: `settings.theme.globalStyles`. Value: () => ({
    'svg > circle': {
      transition: '0.5s all',
    }
  }) 
```
If you have any advice on how to delete this, I'll take it :wink:

**Comment by BenSilver1337** [comment]
Hello!

It seems like you are facing an issue with updating the theme colors in your React app when fetching them from a backend API. You are correct that the theme object might not get updated in the first render, leading to a reference error when using the dynamically created theme colors in your components.

To handle this situation, you can follow these steps:

    Loading State: When fetching the theme colors from the backend API, make sure to set a loading state initially. This will prevent components from trying to access the theme colors before they are available.

    Asynchronous Data Fetching: Fetch the theme colors asynchronously from the backend API, and update the theme object once the data is available. This can be done using the useState hook or a state management library like Redux.

    Conditional Rendering: In your components, check if the theme colors are available before using them. You can use conditional rendering or short-circuit evaluation to ensure that the components do not attempt to access the theme colors until they are updated.
